### DISCUSSION TITLE
`rbenv install 3.1.5` is failed on MacOS 14.4.1

### DISCUSSION THREAD

**Discussion by mustafaergul** [rbenv/ruby-build#2379] [Build failures] [answered]
### **Machine details:**
Apple M1 Pro
macOS 14.4.1

### **Error:**
BUILD FAILED (macOS 14.4.1 on arm64 using ruby-build 20240501)

### **Steps taken to reproduce error:**
- `brew install rbenv ruby-build`
- `rbenv install 3.1.5`

### **Steps taken to resolve error:**
- `xcode-select --install`
- `brew upgrade ruby-build`
- `brew install openssl libffi zlib rbenv readline`
- `brew install apple-gcc42`

I am trying to install ruby 3.1.5 using rbenv but it is failing with the error message as in the file below;
```
/private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:490:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:616:in `block in try_compile'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:563:in `with_werror'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:616:in `try_compile'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:680:in `try_cflags'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:1025:in `block (2 levels) in append_cflags'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:989:in `block in checking_for'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:354:in `block (2 levels) in postpone'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:324:in `open'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:354:in `block in postpone'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:324:in `open'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:350:in `postpone'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:988:in `checking_for'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:1024:in `block in append_cflags'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:1023:in `each'
	from /private/var/folders/xf/057s60sx27s3f1_8jsv9qd940000gn/T/ruby-build.20240502105217.24149.sA8jCT/ruby-3.1.5/lib/mkmf.rb:1023:in `append_cflags'
	from extconf.rb:3:in `<main>'
```

Tried to find something similar in the discussions but specific version case is not found.

[build-log.log](https://github.com/rbenv/ruby-build/files/15185325/build-log.log)

Thanks in advance!

**Comment by mislav** [comment]
Hi, sorry you're having trouble. I see nothing wrong with your setup. What does `xcode-select --install` output?

The `brew install apple-gcc42` step should not be necessary for Ruby. The compilers that ship with the macOS system (obtained through `xcode-select --install`) are enough to build most software.

To check for more information about the failure, these lines of the log file you shared might be a good starting point:
```
6094   │ To see why this extension failed to compile, please check the mkmf.log which can be found here:
6095   │ 
6096   │   /Users/mergul/.rbenv/versions/3.1.5/lib/ruby/gems/3.1.0/extensions/arm64-darwin-23/3.1.0/rbs-2.7.0/mkmf.log
```

**Reply by mustafaergul** [reply]
Thanks for the swift reply ❤️ `xcode-select --install` opens the xcode and installs the tool successfully. 

On the other hand there is such file `/Users/mergul/.rbenv/versions/3.1.5/lib/ruby/gems/3.1.0/extensions/arm64-darwin-23/3.1.0/rbs-2.7.0/mkmf.log`,

**Comment by mislav** [comment, marked as answer]
I found this bug report: https://bugs.ruby-lang.org/issues/19920

It looks like this happens on a macOS system where SIP (system integrity protection) is disabled. By default, [SIP is enabled on macOS](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection).

A workaround might be: `rbenv install 3.1.5 -- --disable-shared`, since it's building the shared library on a SIP-disabled system that triggers the bug.

**Reply by mustafaergul** [reply]
Thanks that is resolved installing the 3.1.5 version 🎸

**Reply by rjhilgefort** [reply]
This worked for me, thanks!

**Reply by Manas8803** [reply]
God send

**Comment by dhanuyapa** [comment]
==> Installed ruby-3.1.5 to /Users/yapabandara/.rbenv/versions/3.1.5

NOTE: to activate this Ruby version as the new default, run: rbenv global 3.1.5
yapabandara@YAPABANDARAs-MacBook-Air ~ % pod --version
zsh: command not found: pod
yapabandara@YAPABANDARAs-MacBook-Air ~ % gem install -n /usr/local/bin ffi cocoapods
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /Library/Ruby/Gems/2.6.0 directory.
yapabandara@YAPABANDARAs-MacBook-Air ~ % echo 'eval "$(rbenv init -)"' >> ~/.zshrc
source ~/.zshrc

yapabandara@YAPABANDARAs-MacBook-Air ~ % gem install ffi cocoapods

ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /Library/Ruby/Gems/2.6.0 directory.
yapabandara@YAPABANDARAs-MacBook-Air ~ %  how to solve this

**Reply by mislav** [reply]
Hi, yours seem to be a problem entirely unrelated to the original problem in this thread. But I will do my best in answering.

First of all, you should not need to install any Ruby version just to install cocoapods. The cocoapods website lists an installation method that uses the system Ruby version that ships with your macOS: https://guides.cocoapods.org/using/getting-started.html#getting-started

However, if you do want to install cocoapods using a newer Ruby version, then you have two options:

1. If your system already has Homebrew, you can use that instead of rbenv: `brew install cocoapods`. This is generally simpler and faster when compared to setting up rbenv for the sole purpose of using `pod`.

2. If you do want to use rbenv for cocoapods, then you are on the right track with what you've shared before, including configuring your zshrc file, but for some reason `gem` still resolves to `/usr/bin/gem` for you instead to `~/.rbenv/shims/gem` which it should after rbenv has been set up. To debug what's happening, open a new terminal window and check `echo $PATH`. Is rbenv's shims directory listed? If so, you can run `rbenv rehash` (just in case) to ensure that `~/.rbenv/shims/gem` gets created.

**Comment by dhanuyapa** [comment]
Thanks
